theano/contrib/mkl/mkl_ndarray.c

```diff
@@ -0,0 +1,1361 @@
+#include <Python.h>
+#include <structmember.h>
+#include <numpy/arrayobject.h>
+#include <stdlib.h>
+#include "mkl_ndarray.h"
+
+
+int MKLNdarray_Check(const PyObject *ob);
+PyObject* MKLNdarray_New(int nd, int typenum);
+int MKLNdarray_CopyFromArray(MKLNdarray *self, PyArrayObject *obj);
+
+
+/*
+ * This function is called in MKLNdarray_dealloc.
+ *
+ * Release all allocated buffer and layout in self.
+ *
+ * If private_layout is a reference from another ndarray,
+ * do not free it in this ndarray.
+ *
+ */
+static int
+MKLNdarray_uninit(MKLNdarray *self) {
+    int rval = 0;
+    if (self->dtype == MKL_FLOAT64) {  // for float64
+        if (self->private_data) {
+            rval = dnnReleaseBuffer_F64(self->private_data);
+
+            if (rval != E_SUCCESS) {
+                PyErr_Format(PyExc_RuntimeError,
+                             "MKLNdarray_uninit: fail to release data: %d, line: %d",
+                             rval, __LINE__);
+            }
+            self->private_data = NULL;
+        }
+
+        if ((self->base == NULL) && self->private_layout) {
+            rval = dnnLayoutDelete_F64(self->private_layout);
+
+            if (rval != E_SUCCESS) {
+                PyErr_Format(PyExc_RuntimeError,
+                             "MKLNdarray_uninit: fail to release layout: %d, line: %d",
+                             rval, __LINE__);
+            }
+            self->private_layout = NULL;
+        }
+
+        if (self->private_workspace) {
+            rval = dnnReleaseBuffer_F64(self->private_workspace);
+
+            if (rval != E_SUCCESS) {
+                PyErr_Format(PyExc_RuntimeError,
+                             "MKLNdarray_uninit: fail to release workspace: %d, line: %d",
+                             rval, __LINE__);
+            }
+            self->private_workspace = NULL;
+        }
+    } else {  // for float32
+        if (self->private_data) {
+            rval = dnnReleaseBuffer_F32(self->private_data);
+
+            if (rval != E_SUCCESS) {
+                PyErr_Format(PyExc_RuntimeError,
+                             "MKLNdarray_uninit: fail to release data: %d, line: %d",
+                             rval, __LINE__);
+            }
+            self->private_data = NULL;
+        }
+
+        if ((self->base == NULL) && self->private_layout) {
+            rval = dnnLayoutDelete_F32(self->private_layout);
+
+            if (rval != E_SUCCESS) {
+                PyErr_Format(PyExc_RuntimeError,
+                             "MKLNdarray_uninit: fail to release layout: %d, line: %d",
+                             rval, __LINE__);
+            }
+            self->private_layout = NULL;
+        }
+
+        if (self->private_workspace) {
+            rval = dnnReleaseBuffer_F32(self->private_workspace);
```

Comment:
Regarding the dnn function names, it was a legacy design in MKL that including dtype in it. 

So with such fixed APIs, we can't avoid duplicated codes, no matter we write them in this way or create a new header file which combines the two functions into a wrapper by using a parameter "dtype" to distinguish different data types.

So I suggest to keep the code as it was for now, and align to MKL when it evolves.